File: backend/app/models/ticket.py
from __future__ import annotations

from sqlalchemy import ForeignKey, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from backend.app.models.base import Base, TimestampMixin, generate_id


class Ticket(TimestampMixin, Base):
    __tablename__ = "tickets"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=generate_id)
    product_id: Mapped[str | None] = mapped_column(
        ForeignKey("products.id", ondelete="SET NULL"),
        nullable=True,
    )
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False)
    fault_category: Mapped[str] = mapped_column(String(100), nullable=False, default="")
    priority: Mapped[str] = mapped_column(String(32), nullable=False, default="medium")
    status: Mapped[str] = mapped_column(String(32), nullable=False, default="open")
    solution: Mapped[str] = mapped_column(Text, nullable=False, default="")
    resolution_notes: Mapped[str] = mapped_column(Text, nullable=False, default="")

    product = relationship("Product")

File: frontend/src/types/ticket.ts
export interface Ticket {
  id: string
  product_id: string | null
  title: string
  description: string
  fault_category: string
  priority: 'low' | 'medium' | 'high' | 'urgent'
  status: 'open' | 'in_progress' | 'resolved' | 'closed'
  solution: string
  resolution_notes: string
  created_at: string
  updated_at: string
}

export interface SimilarTicket {
  ticket: Ticket
  similarity: number
}
